Standardize/Fix behavior of JsLookup#apply and JsLookup#\

lihaoyi opened the pull request:
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Typesafe CLA](https://www.typesafe.com/contribute/cla)?
* [x] Have you [squashed your commits](https://www.playframework.com/documentation/latest/WorkingWithGit#Squashing-commits)?
* [x] Have you added copyright headers to new files?
* [x] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?

## Fixes

I don't see an open issue for this

## Purpose

- Fix behavior of `JsLookup#apply(index: Int)` to properly throw exceptions, as is already described in the readme.

- Add a `JsLookup#apply(fieldName: String)` which similarly throws exceptions if the key is not found, for consistency with `JsLookup#apply(index: Int)` and `JsLookup#\(index: Int)` and `JsLookup#\(fieldName: String)`

- Convert an internal callsite from `JsLookup#apply(index: Int)` to `JsLookup#\(index: Int)`, since that callsite seems to want the don't-throw-exception behavior

- Document behavior of existing `JsLookup#\(index: Int)` and new `JsLookup#apply(fieldName: String)` in the readme

- Add tests for:

    - The existing `JsLookup#apply(index: Int)` to ensure it throws exceptions as is described in the readme,
    - The new `JsLookup#apply(fieldName: String)` operation

## Background Context

- `JsLookup#apply(index: Int)` was a simple alias for `JsLookup#\(index: Int)`, which meant it wasn't throwing which is what the existing readme said it should be doing

- If we have both exception-throwing and non-exception-throwing ways of looking up `JsArrays` by `index: Int`, it makes sense to have the same two ways of looking up `JsObjects` by `fieldName: String`

I haven't discussed this with anyone, but given the existing code and the existing behavior the readme describes, this seems like the right thing to do. Tests seem to pass, though the old incorrect-according-to-readme behavior of `JsLookup#apply(index: Int)` didn't appear to be tested anywhere so the change in semantics didn't break anything until I added my own tests.

Let me know what you think...

## References

lihaoyi commented:
Test failure looks like a flake

gmethvin commented:
@lihaoyi For the examples we probably want to update the test here instead of creating a new test: https://github.com/playframework/play-json/blob/master/docs/manual/working/scalaGuide/main/json/code/ScalaJsonSpec.scala#L14

The compiled docs are here: https://www.playframework.com/documentation/2.6.x/ScalaJson, and the markdown source is here: https://github.com/playframework/play-json/blob/master/docs/manual/working/scalaGuide/main/json/ScalaJson.md

I'm not sure what to do for the README. I wanted to have a simple getting started guide in this repo, with the more complex examples in the compiled Play documentation.

gmethvin commented:
The idea is that `JsLookup` is a "safe" lookup. So it should return a `JsDefined`/`JsUndefined` depending on whether the key exists. This is effectively the way it's always been in play-json, though `JsUndefined` used to be included in the `JsValue` type.

It seems like the `JsLookup.apply` should be deprecated, since it's easily confused for `JsArray.apply`, and we suggest using `\` instead.

lihaoyi commented:
Thanks for the pointer to the tests in the `doc/` folder, I've updated them and got rid of the redundant tests I added.

I'm not sure what `JsArray#apply` method you are referring to? I don't see any in the codebase.

From a API-design perspective, I think that it makes sense to have an "unsafe" lookup easily available: we already have `\` to be the "safe" lookup, and my experience there are often use-cases where you just want to pull out data you know exists, and if it doesn't exist you want your program to crash with a (hopefully detailed) error message.

Perhaps that occurs more in small scripts than in large web applications, but I think it would be nice to be able to use play-json in small scripts rather than have to pull in a totally separate library like uPickle, especially if the necessary changes are so simple/superficial.

From an understandability point of view, I think people would not be confused by the difference between `\` and `.apply`: `.apply` exists in plain Scala `Map`s, `Array`s, `Vector`s, etc. and in all cases does the same thing this diff would make it do here: it looks up some key/index, and throws if it's not found. I think adding similar functionality to play-json would go a long way to help make the library easy to use in less-strict contexts like scripts, scrapers, dev-ops automation or the interactive REPL where I would just be calling `.get` everywhere anyway.

I understand that this is a non-trivial change in the API, so I don't expect you to merge it instantly, but I think it is a relatively-simple change that shouldn't confuse people and will make play-json much more convenient to use directly in less-strict contexts

gmethvin commented:
> I'm not sure what JsArray#apply method you are referring to? I don't see any in the codebase.

Sorry, for some reason I thought there was an unsafe `JsArray#apply` method already.

I'm okay with this change as long as we document the fact that both `apply` methods are "unsafe". When people upgrade, existing code will fail to compile, so we will just want to add something to the Play 2.6 migration guide to help people upgrading.

gmethvin commented on docs/manual/working/scalaGuide/main/json/ScalaJson.md:
```diff
@@ -101,22 +101,26 @@ You can traverse a `JsValue` structure and extract specific values. The syntax a
 
 @[traverse-simple-path](code/ScalaJsonSpec.scala)
 
+In this example we are using `.get`, which retrieves the value out of the resulting `JsResult` and throws an exception if it was not found. You can also use the `JsResult` directly to exceptions, or use the Direct lookup `.apply` method below if you want a missing key or index to throw.
```
Here's a suggestion to clarify (note that `JsResult` is the result of parsing JSON, while `JsLookupResult` is the result of a lookup):

The `\` operator returns a `JsLookupResult`, which is either `JsDefined` or `JsUndefined`. You can chain multiple `\` operators, and the result will be `JsUndefined` if any intermediate value cannot be found. Calling `get` on a `JsLookupResult` attempts to get the value if it is defined and throws an exception if it does not exist.

You can also use the `apply` method to get a field in an object or index in an array. Like `get`, this method will throw an exception if the value does not exist.

lihaoyi commented:
I have updated the docs as you suggested.

I assume the play 2.6 migration guide would be in a separate repo somewhere else?

gmethvin commented on docs/manual/working/scalaGuide/main/json/ScalaJson.md:
```diff
@@ -101,22 +101,29 @@ You can traverse a `JsValue` structure and extract specific values. The syntax a
+### Direct lookup
 
-You can retrieve a value in a `JsArray` using an apply operator with the index number.
+You can retrieve a value in a `JsArray` or `JsObject` using an `.apply` operator, which is identical to the Simple path `\` operator except it returns the value directly (rather than wrapping it in a `JsResult`) throws an exception if the index or key is not found:
```
Should be:

> which is identical to the *s*imple path `\` operator except it returns the value directly (rather than wrapping it in a *`JsLookupResult`*)* and* throws an exception if the index or key is not found

(changes italicised)

lihaoyi commented:
Oops, added back the missing `and`

gmethvin commented:
Yeah, the migration guide is here: https://github.com/playframework/playframework/blob/master/documentation/manual/releases/release26/migration26/Migration26.md

I created an issue: playframework/playframework#7348. If you want you can resolve it, otherwise I'll get to it when I make another sweep through the migration docs in the next few weeks.

gmethvin commented on docs/manual/working/scalaGuide/main/json/ScalaJson.md:
```diff
@@ -101,22 +101,29 @@ You can traverse a `JsValue` structure and extract specific values. The syntax a
+### Direct lookup
 
-You can retrieve a value in a `JsArray` using an apply operator with the index number.
+You can retrieve a value in a `JsArray` or `JsObject` using an `.apply` operator, which is identical to the Simple path `\` operator except it returns the value directly (rather than wrapping it in a `JsResult`) and throws an exception if the index or key is not found:
```
Sorry should have been more clear. It's JsLookupResult not JsResult.

lihaoyi commented:
Changed `JsResult` -> `JsLookupResult`

lihaoyi commented:
Thank @gmethvin!
